tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c

```diff
@@ -0,0 +1,46 @@
+/* FreeRTOS includes. */
+#include "FreeRTOS.h"
+#include "queue.h"
+
+/* FreeRTOS+TCP includes. */
+#include "FreeRTOS_IP.h"
+#include "FreeRTOS_IP_Private.h"
+#include "FreeRTOS_TCP_IP.h"
+
+/* This proof assumes that prvTCPPrepareSend and prvTCPReturnPacket are correct.
+These functions are proved to be correct separately. */
+
+
+BaseType_t publicTCPHandleState( FreeRTOS_Socket_t *pxSocket, NetworkBufferDescriptor_t **ppxNetworkBuffer );
+
+/* Implementation of safe malloc */
+void *safeMalloc(size_t xWantedSize) {
+	if(xWantedSize == 0) {
+		return NULL;
+	}
+	uint8_t byte;
+	return byte ? malloc(xWantedSize) : NULL;
+}
+
+
+void harness() {
+	FreeRTOS_Socket_t *pxSocket = safeMalloc(sizeof(FreeRTOS_Socket_t));
+	if (pxSocket) {
+		pxSocket->u.xTCP.rxStream = safeMalloc(sizeof(StreamBuffer_t));
+		pxSocket->u.xTCP.txStream = safeMalloc(sizeof(StreamBuffer_t));
+		pxSocket->u.xTCP.pxPeerSocket = safeMalloc(sizeof(FreeRTOS_Socket_t));
+		__CPROVER_assume(pxSocket->u.xTCP.xTCPWindow.ucOptionLength == ipSIZE_TCP_OPTIONS/sizeof(uint32_t));
+		__CPROVER_assume(pxSocket->u.xTCP.uxRxWinSize >= 0 && pxSocket->u.xTCP.uxRxWinSize <= 0xfffful);
+		__CPROVER_assume(pxSocket->u.xTCP.usInitMSS == 2);
+	}
+	
+	NetworkBufferDescriptor_t *pxNetworkBuffer = safeMalloc(sizeof(NetworkBufferDescriptor_t));
+	if(pxNetworkBuffer) {
+		pxNetworkBuffer->pucEthernetBuffer = safeMalloc(sizeof(TCPPacket_t));
+	}
+	if (pxSocket && pxNetworkBuffer && pxNetworkBuffer->pucEthernetBuffer) {
+		publicTCPHandleState(pxSocket, &pxNetworkBuffer);
```

Comment:
This returns an error code.  What properties do we expect to be true if error is returned - perhaps that the input was unmodified? What do we expect to be true if error was not returned.  i.e. surely this function has post-conditions? What are they? Can we check them?

Reply:
I have not done these checks before. I discussed with Mark and we decided to proof memory safety first. After we finish it we will revisit these proofs and implement these checks.